docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+A suggestion from Chris Lattner provides a path to avoiding that by having clang synthesise wrapper functions.
+
+For example, consider some Verona code that wants to call `llvm::IRBuilder<>::CreateRetVoid`.
+This is a method in a templated class, so the template will need to be instantiated before it can be called and the caller then needs to be aware of C++ ABI details.
+This, like every other problem in computer science, can be addressed by a level of indirection, using clang to synthesise a function roughly equivalent to this :
+
+```c++
+__attribute__((used,always_inline))
+extern "C" llvm::ReturnInst *irbuilder_createretvoid(llvm::IRBuilder<> *b)
+{
+        return b->CreateRetVoid();
+}
+```
+
+This takes and returns all values as pointers and so the LLVM lowering is predictable as a function that takes and returns only pointers.
+For complex return values, the caller will need to allocate space and pass a pointer to the return.
+Clang generates IR somewhat like the following for this function.
+
+```llvm
+define dso_local %"class.llvm::ReturnInst"* @irbuilder_createretvoid(%"class.llvm::IRBuilder"*) #0 {
+  %2 = alloca %"class.llvm::IRBuilder"*, align 8
+  store %"class.llvm::IRBuilder"* %0, %"class.llvm::IRBuilder"** %2, align 8
+  %3 = load %"class.llvm::IRBuilder"*, %"class.llvm::IRBuilder"** %2, align 8
+  %4 = bitcast %"class.llvm::IRBuilder"* %3 to %"class.llvm::IRBuilderBase"*
+  %5 = call %"class.llvm::ReturnInst"* @_ZN4llvm13IRBuilderBase13CreateRetVoidEv(%"class.llvm::IRBuilderBase"* %4)
+  ret %"class.llvm::ReturnInst"* %5
+}
+```
+
+Creating this forced clang to instantiate a lot of templates.
+Fortunately, [Hal Finkel's C++ JIT prototype](https://github.com/hfinkel/llvm-project-cxxjit) demonstrated that this is possible without changes to the clang AST.
+
+This generated function is then trivial for the Verona compiler to call (all details of the C/C++ ABI are abstracted) and can then be inlined during the optimisation phase if desirable.
+More interestingly, LLVM will inline the `IRBuilder<>::CreateRetVoid` call into this function.
+The Verona compiler can then examine the result and determine if it is safe to inline at the caller without breaking the sandboxing invariants.
+For example, a function that ends up flattened can be made safe simply by enforcing the sandbox bounds checks on every LVM IR load and store instruction.
+Functions that do a lot of work, in contrast, may require a full domain switch.
+
+Exactly the same technique can be applied to field accessors.
+Getting and setting a field in a C or C++ POD `struct` is easy in the common cases, but may involve some complex ABI interactions for features such as bitfields in structures where some fields have custom alignment.
+To access these fields, the Verona compiler would simply make clang synthesize a 'get' and 'set' function for each field.
+This can then be inlined (with correct bounds checks applied).
```

Comment:
Would this be allowed to violate encapsulation? If not, would the compiler be required to understand the target language's properties?

Reply:
Good question.  I would expect it not to be allowed to violate encapsulation, so the clang code would need to perform the relevant access checks.  This would involve reusing / copying and pasting some sema logic.